pkg/streamingpromql/operators/instant_vector_to_scalar.go

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package operators
+
+import (
+	"context"
+	"math"
+
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser/posrange"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/limiting"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+// InstantVectorToScalar is an operator that implements the scalar() function.
+type InstantVectorToScalar struct {
+	Inner                    types.InstantVectorOperator
+	Start                    int64 // Milliseconds since Unix epoch
+	End                      int64 // Milliseconds since Unix epoch
+	Interval                 int64 // In milliseconds
+	MemoryConsumptionTracker *limiting.MemoryConsumptionTracker
+
+	expressionPosition posrange.PositionRange
+}
+
+var _ types.ScalarOperator = &InstantVectorToScalar{}
+
+func NewInstantVectorToScalar(
+	inner types.InstantVectorOperator,
+	start int64,
+	end int64,
+	interval int64,
+	memoryConsumptionTracker *limiting.MemoryConsumptionTracker,
+	expressionPosition posrange.PositionRange,
+) *InstantVectorToScalar {
+	return &InstantVectorToScalar{
+		Inner:                    inner,
+		Start:                    start,
+		End:                      end,
+		Interval:                 interval,
+		MemoryConsumptionTracker: memoryConsumptionTracker,
+		expressionPosition:       expressionPosition,
+	}
+}
+
+func (i *InstantVectorToScalar) GetValues(ctx context.Context) (types.ScalarData, error) {
+	seriesCount, err := i.getInnerSeriesCount(ctx)
+	if err != nil {
+		return types.ScalarData{}, err
+	}
+
+	stepCount := stepCount(i.Start, i.End, i.Interval)
+	seenPoint, err := types.BoolSlicePool.Get(stepCount, i.MemoryConsumptionTracker)
+	if err != nil {
+		return types.ScalarData{}, err
+	}
+
+	defer types.BoolSlicePool.Put(seenPoint, i.MemoryConsumptionTracker)
+	seenPoint = seenPoint[:stepCount]
+
+	output, err := types.FPointSlicePool.Get(stepCount, i.MemoryConsumptionTracker)
+	if err != nil {
+		return types.ScalarData{}, err
+	}
+
+	for t := i.Start; t <= i.End; t += i.Interval {
+		output = append(output, promql.FPoint{
+			T: t,
+			F: math.NaN(),
+		})
+	}
+
+	for seriesIdx := 0; seriesIdx < seriesCount; seriesIdx++ {
```

Comment:
Why do this instead of looping until we see EOS?

Reply:
I did it this way so we could detect a situation where the inner operator said it would give X series but only returned Y series.

Granted, that should never happen, but if it did, that'd be weird and we'd want to fail the query rather than silently continue.